src/inspect/mod.rs

```diff
@@ -2,238 +2,355 @@
-            TypeHintExpr::ModuleAttribute { module, attr } => {
-                module.fmt(f)?;
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Constant { value } => match value {
+                PyStaticConstant::None => f.write_str("None"),
+            },
+            Self::Name { id, .. } => f.write_str(id),
+            Self::Attribute { value, attr } => {
+                value.fmt(f)?;
                 f.write_str(".")?;
-                attr.fmt(f)
+                f.write_str(attr)
+            }
+            Self::BinOp { left, op, right } => {
+                left.fmt(f)?;
+                f.write_char(' ')?;
+                f.write_char(match op {
+                    PyStaticOperator::BitOr => '|',
+                })?;
+                f.write_char(' ')?;
+                right.fmt(f)
             }
-            TypeHintExpr::Union { elts } => {
-                for (i, elt) in elts.iter().enumerate() {
-                    if i > 0 {
-                        f.write_str(" | ")?;
-                    }
-                    elt.fmt(f)?;
+            Self::Tuple { elts } => {
+                f.write_char('(')?;
+                fmt_elements(elts, f)?;
+                if elts.len() == 1 {
+                    f.write_char(',')?;
                 }
-                Ok(())
+                f.write_char(')')
             }
-            TypeHintExpr::Subscript { value, slice } => {
+            Self::List { elts } => {
+                f.write_char('[')?;
+                fmt_elements(elts, f)?;
+                f.write_char(']')
+            }
+            Self::Subscript { value, slice } => {
                 value.fmt(f)?;
                 f.write_str("[")?;
-                for (i, elt) in slice.iter().enumerate() {
-                    if i > 0 {
-                        f.write_str(", ")?;
-                    }
-                    elt.fmt(f)?;
+                if let PyStaticExpr::Tuple { elts } = slice {
+                    // We don't display the tuple parentheses
+                    fmt_elements(elts, f)?;
+                } else {
+                    slice.fmt(f)?;
                 }
                 f.write_str("]")
             }
         }
     }
 }
 
+/// A PyO3 extension to the Python AST to know more about [`PyStaticExpr::Name`].
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub enum PyStaticNameKind {
+    /// A local name, relative to the current module
+    Local,
+    /// A global name, can be a module like `datetime`, a builtin like `int`...
+    Global,
+}
+
+/// A PyO3 extension to the Python AST to know more about [`PyStaticExpr::Constant`].
+///
+/// This enables advanced features like escaping.
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub enum PyStaticConstant {
+    /// None
+    None,
+    // TODO: add Bool(bool), String(&'static str)... (is useful for Literal["foo", "bar"] types)
```

Comment:
`String(&'static str)` would also be interesting because it can be used as an escape hatch for forward-reference style annotations for things we can't parse.

e.g. we could have `"tuple[()]"` already as a type hint until we can support it natively (with a `PyStaticCall` ast node, I guess).